mlir/lib/Bindings/Python/IRAttributes.cpp

```diff
@@ -205,14 +316,13 @@ class PyDenseArrayAttribute : public PyConcreteAttribute<DerivedT> {
     EltTy dunderNext() {
       // Throw if the index has reached the end.
       if (nextIndex >= mlirDenseArrayGetNumElements(attr.get()))
-        throw py::stop_iteration();
+        throw nb::stop_iteration();
```

Comment:
I wish this hadn't been cargo culted from pybind. Exceptions as control flow perform badly. In hot paths on other projects, I did something lower level vs throwing, but better to just switch in kind for now.

Reply:
What's the lower level alternative? AFAIK throwing `StopIteration` is only way to exit a generator?

Reply:
It turns out there is an example later in this PR (for generic ir iterators). The C iteration protocol doesn't require c++ exceptions, which are the problems. It should have been implemented in PB/NB to be based on a return value vs an exception. ABC then that would just result in setting the python error status directly, which is cheap. 

Design flaw in pybind that propagated to nanobind.

Reply:
Yes, I agree. C++ exceptions are slow and shouldn't be used in cases that happen frequently. Not new or changed in this PR, though.

We had a similar problem in JAX for a different exception (`KeyError`) and method (`__getattr__`). It's not hard to attach a classic C API method to a nanobind or pybind class if it turns out to matter for performance, e.g., see: https://github.com/openxla/xla/blob/f652407e382f7eb2fe58ae0eec6b8edd31e68f63/xla/python/py_device.cc#L314

In fact we do this trick in both directions. We also attach nanobind-defined methods to classes defined using the low level C APIs.